pkg/auth/jwt.go

```diff
@@ -68,25 +80,41 @@ func retrieveJwtToken(ctor RequestCtor, doer RequestDoer) (*expiringToken, error
 	return &expiringToken{token, expiration}, nil
 }
 
-// Augments the given doer with an authorization header carrying the
-// given token. The token gets refreshed when it becomes stale.
-func atachJwtToken(ctor RequestCtor, doer RequestDoer, token *expiringToken) RequestDoer {
-	lock := sync.Mutex{}
-	return func(req *http.Request) (*http.Response, error) {
-		lock.Lock()
-		if token.expiration.Before(time.Now()) {
-			// Keeping the lock while making http request is dubious, but
-			// making multiple concurrent requests is not any better.
-			t, err := retrieveJwtToken(ctor, doer)
-			if err != nil {
-				// Can't get a new token, so returning the error preemptively.
-				lock.Unlock()
-				return nil, err
-			}
-			*token = *t
+// TokenCache keeps the tokens for reuse by Get.
+type TokenCache struct {
+	lock  sync.Mutex
+	token *expiringToken
+	ctor  requestCtor
+	doer  requestDoer
+}
+
+// MakeCache creates a new cache or returns an error if tokens aren't
+// available.
+func MakeCache(ctor func(method, url string, body io.Reader) (*http.Request, error),
+	doer func(req *http.Request) (*http.Response, error)) (*TokenCache, error) {
+	token, err := retrieveJwtToken(ctor, doer)
+	if err != nil {
+		return nil, err
+	}
+	return &TokenCache{sync.Mutex{}, token, ctor, doer}, nil
+}
+
+// Get returns a potentially cached value of the token or renews as
+// necessary. The now parameter provides the current time for cache
+// expiration. The returned value is suitable for Authorization header
+// and syz-hub Key requests.
+func (cache *TokenCache) Get(now time.Time) (string, error) {
+	cache.lock.Lock()
+	defer cache.lock.Unlock()
+	if cache.token.expiration.Before(now) {
```

Comment:
If we do exact match we can fail requests around the expiration time with significant probability.
I think it's better to do something like `cache.token.expiration.Sub(now) < time.Minute` (assuming token is valid for >> time.Minute).

Reply:
Good call, done. Though we know that there's always a sufficiently long tail of stragglers to make anything but unlimited retry insufficient. Stories of data centers recovering from weeks of disconnect are a solid part of SRE lore.